client/http/encoding/json/json.go

```diff
@@ -0,0 +1,101 @@
+// Package json provides helper functions to handle requests and responses.
+package json
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+
+	"github.com/beatlabs/patron/encoding"
+	"github.com/beatlabs/patron/encoding/json"
+	"github.com/beatlabs/patron/log"
+)
+
+// NewRequest creates a request and encodes the body and set's the appropriate headers.
+func NewRequest(ctx context.Context, method string, url string, payload interface{}) (*http.Request, error) {
+	buf, err := json.Encode(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, method, url, bytes.NewBuffer(buf))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set(encoding.ContentTypeHeader, json.Type)
+	req.Header.Set(encoding.ContentLengthHeader, strconv.FormatInt(int64(len(buf)), 10))
+	req.Header.Set(encoding.AcceptHeader, json.Type)
+
+	return req, nil
+}
+
+// FromResponse checks for valid headers and decodes th payload.
+func FromResponse(ctx context.Context, rsp *http.Response, payload interface{}) error {
+	err := validatingContentTypeHeader(rsp)
+	if err != nil {
+		return err
+	}
+
+	buf, err := io.ReadAll(rsp.Body)
+	defer func() {
+		err := rsp.Body.Close()
+		if err != nil {
+			log.FromContext(ctx).Errorf("failed to close response body: %v", err)
+		}
+	}()
+	if err != nil {
+		return err
+	}
+
+	err = validateContentLength(rsp, len(buf))
+	if err != nil {
+		log.FromContext(ctx).Warn(err)
```

Comment:
This is an open question: We are checking that the content length of the header and the payload are correct. In case it is not I am only logging but do not fail.
Question: Should we fail? Keep in mind that decoding worked.

Reply:
Thanks for the insight.
I will leave it as a warning log since the problem might be much bigger.
If we use gzip the length needs to be the compressed length and not the original one.
Let's see how it goes with logging first.